### ISSUE TITLE
Expose clipboard formats in Mac and Linux

### ISSUE THREAD

**Issue by LexiconCode** [dictation-toolbox/dragonfly#226]
Currently pyperclip does not support exposing the format of contents in the clipboard besides plaintext in Mac and Linux.

### Mac OS
For Mac OS we can dump the formats using `richxerox` module [details](https://stackoverflow.com/questions/32959928/accessing-alternate-clipboard-formats-from-python) 

### Linux
xclip may be the best option
https://askubuntu.com/questions/705620/xclip-vs-xsel
https://stackoverflow.com/questions/3571179/how-does-x11-clipboard-handle-multiple-data-formats
https://stackoverflow.com/questions/3261379/getting-html-source-or-rich-text-from-the-x-clipboard

### Other resources
A few repositories to keep an eye on in the future.
https://pypi.org/project/jaraco.clipboard/
https://github.com/OzymandiasTheGreat/klembord
https://github.com/bugaevc/wl-clipboard

**Comment by Danesprite** [comment]
Thanks for this detailed post! I like the idea of exposing other clipboard formats for each platform. I will have a look into your links and figure out the best way forward.

**Comment by LexiconCode** [comment]
Here's some work I've done for Linux.
- `get_clipboard_formats` uses xclip to grab the contents of the clipboard.
- `enum_files_from_clipboard` clipboard contents that begins with `file://`. Then generates a list by processing into unverified absolute file/folder paths.
- `get_clipboard_files` calls `enum_files_from_clipboard` based on available contents clipboard mime. Then validates file/folder paths effectively emulating `win32clipboard.CF_UNICODETEXT`

```
from locale import getpreferredencoding
import sys, os, subprocess
from six import binary_type
try:
    from urllib import unquote
except ImportError:
    from urllib.parse import unquote

def get_clipboard_formats():
    '''
    Return list of all data formats currently in the clipboard
    '''
    formats = []
    if sys.platform.startswith('linux'):
        encoding = getpreferredencoding()
        com = ["xclip", "-o", "-t", "TARGETS"]
        try:
            p = subprocess.Popen(com,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                stdin=subprocess.PIPE,
                                )
            for line in iter(p.stdout.readline, b''):
                if isinstance(line, binary_type):
                    line = line.decode(encoding)
                    formats.append(line.strip())
        except Exception as e:
            print("Exception from starting subprocess {0}: " "{1}".format(com, e))

    if sys.platform.startswith('win'):
        import win32clipboard  # pylint: disable=import-error
        f = win32clipboard.EnumClipboardFormats(0)
        while f:
            formats.append(f)
            f = win32clipboard.EnumClipboardFormats(f)    

    if not formats:
        print("get_clipboard_formats: formats are {}: Not implemented".format(formats))
    else:
        return formats

def enum_files_from_clipboard(target):
    '''
    Generates absolute paths from clipboard 
    Returns file/dir based on target mime type
    '''
    formats = []
    if sys.platform.startswith('linux'):
        encoding = getpreferredencoding()
        com = ["xclip", "-selection", "clipboard","-o", target]
        try:
            p = subprocess.Popen(com,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                stdin=subprocess.PIPE,
                                )
            for line in iter(p.stdout.readline, b''):
                if isinstance(line, binary_type):
                    line = line.decode(encoding).strip()
                    if line.startswith("file://"):
                        formats.append(unquote(line.replace("file://", "")))
            #print(formats)
            return formats
        except Exception as e:
            print("Exception from starting subprocess {0}: " "{1}".format(com, e))

def get_clipboard_files(folders=False):
    '''
    Enumerate clipboard content and return files either directly copied or
    highlighted path copied
    '''
    files = None
    if sys.platform.startswith('win'):
        import win32clipboard  # pylint: disable=import-error
        win32clipboard.OpenClipboard()
        f = get_clipboard_formats()
        if win32clipboard.CF_HDROP in f:
            files = win32clipboard.GetClipboardData(win32clipboard.CF_HDROP)
        elif win32clipboard.CF_UNICODETEXT in f:
            files = [win32clipboard.GetClipboardData(win32clipboard.CF_UNICODETEXT)]
        elif win32clipboard.CF_TEXT in f:
            files = [win32clipboard.GetClipboardData(win32clipboard.CF_TEXT)]
        elif win32clipboard.CF_OEMTEXT in f:
            files = [win32clipboard.GetClipboardData(win32clipboard.CF_OEMTEXT)]
        if folders:
            files = [f for f in files if os.path.isdir(f)] if files else None
        else:
            files = [f for f in files if os.path.isfile(f)] if files else None
        win32clipboard.CloseClipboard()
        return files

    if sys.platform.startswith('linux'):
        f = get_clipboard_formats()
        if "UTF8_STRING" in f:
            files = enum_files_from_clipboard("UTF8_STRING")
        elif "TEXT" in f:
            files = enum_files_from_clipboard("TEXT")
        elif "text/plain" in f:
            files = enum_files_from_clipboard("text/plain")
       if folders:
            files = [f for f in files if os.path.isdir(str(f))] if files else None
        else:
            files = [f for f in files if os.path.isfile(str(f))] if files else None
```

**Comment by Danesprite** [comment]
Neat! What desktop environment did you test with on Linux? Unfortunately, copied files/folders aren't showing up for me on Xfce. This is because the default file manager (Thunar) doesn't prefix copied files with `"file://"`. I wonder if there is some sort of X11 standard for clipboard formats.

**Comment by LexiconCode** [comment]
> Neat! What desktop environment did you test with on Linux? Unfortunately, copied files/folders aren't showing up for me on Xfce. This is because the default file manager (Thunar) doesn't prefix copied files with `"file://"`. Or maybe my window manager is at fault? Not sure.. I wonder if there is some sort of X11 standard for clipboard formats.

Oh that's too bad. My Ubuntu/Pop_os windows manager gnome. How does Thunar represent files/folders in the clipboard and what does get_clipboard_formats return? If it's absolute file/folder we can simply blindly check if path is valid. There are standard mime type but as far as I can tell there's no standard files/folder.

**Comment by Danesprite** [comment]
Yes, Thunar represents copied files and folders as simple absolute paths. I have got this working properly by changing your `get_clipboard_formats()` and `enum_files_from_clipboard()` functions:

```Python
def get_clipboard_formats():
    '''
    Return list of all data formats currently in the clipboard
    '''
    formats = []
    if sys.platform.startswith('linux'):
        encoding = getpreferredencoding()
        com = ["xclip", "-o", "-t", "TARGETS"]
        try:
            p = subprocess.Popen(com,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                stdin=subprocess.PIPE,
                                )
            for line in iter(p.stdout.readline, b''):
                if isinstance(line, binary_type):
                    line = line.decode(encoding)
                formats.append(line.strip())
        except Exception as e:
            print("Exception from starting subprocess {0}: " "{1}".format(com, e))

    if sys.platform.startswith('win'):
        import win32clipboard  # pylint: disable=import-error
        f = win32clipboard.EnumClipboardFormats(0)
        while f:
            formats.append(f)
            f = win32clipboard.EnumClipboardFormats(f)    

    if not formats:
        print("get_clipboard_formats: formats are {}: Not implemented".format(formats))
    else:
        return formats

def enum_files_from_clipboard(target):
    '''
    Generates absolute paths from clipboard 
    Returns file/dir based on target mime type
    '''
    formats = []
    if sys.platform.startswith('linux'):
        encoding = getpreferredencoding()
        com = ["xclip", "-selection", "clipboard","-o", target]
        try:
            p = subprocess.Popen(com,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                stdin=subprocess.PIPE,
                                )
            for line in iter(p.stdout.readline, b''):
                if isinstance(line, binary_type):
                    line = line.decode(encoding).strip()
                if line.startswith("file://"):
                    line = line.replace("file://", "")
                formats.append(unquote(line))
            #print(formats)
            return formats
        except Exception as e:
            print("Exception from starting subprocess {0}: " "{1}".format(com, e))

```

`get_clipboard_formats()` with my changes returns the following:

```Python
[u'TIMESTAMP', u'TARGETS', u'MULTIPLE', u'UTF8_STRING', u'COMPOUND_TEXT', u'TEXT', u'STRING', u'text/plain;charset=utf-8', u'text/plain']
```

Sorry about the late reply BTW.

**Comment by LexiconCode** [comment]
No problem, thanks! 

@kendonB you may be interested in this.

**Comment by Danesprite** [comment]
So is the idea to add a new Dragonfly `Clipboard` class implementation using `xsel` for X11 and adjust the Windows implementation?

**Comment by LexiconCode** [comment]
> 
> 
> So is the idea to add a new Dragonfly `Clipboard` class implementation using `xsel` for X11 and adjust the Windows implementation?


Yes that could work that way if you prefer.

**Comment by Danesprite** [comment]
Okay then. I can test with Mac OS, but I am not familiar enough with Apple's APIs to implement this. Do you want me to write an `X11Clipboard` class for `xsel` using your code? Not sure when I will get to it.

**Comment by LexiconCode** [comment]
> Do you want me to write an `X11Clipboard` class for `xsel` using your code? Not sure when I will get to it.

Yes that would be nice but I considerate it a low priority so take your time.

**Comment by Danesprite** [comment]
Okay then :+1:

**Comment by Danesprite** [comment]
I think this issue should be closed once the next version is released. X11 is done now and it should be fairly trivial to further expose clipboard formats in the future, if necessary. It is also much easier to implement `Clipboard` classes for other interfaces or platforms, e.g. for Wayland.

@LexiconCode Thanks very much for your above initial code. It was instrumental in getting the new classes written. :-)

I'm not really willing to put much time into this for macOS. The [`pbpaste`](https://ss64.com/osx/pbpaste.html) command-line program for getting data from the clipboard doesn't return full file paths for copied files/folders, so it can't be used to implement the `format_hdrop` clipboard format. Presumably, this can be done using the AppKit API (see [Documents, Data, and Pasteboard](https://developer.apple.com/documentation/appkit/documents_data_and_pasteboard)).

As for Rich Text Format (RTF), I think this is out of the scope of this project. In any case, looking at the manual page for `pbpaste`, it would probably be simpler for those interested to just use that program directly.

**Comment by LexiconCode** [comment]
Sounds reasonable!
